inspirehep/records/api/base.py

```diff
@@ -166,3 +179,228 @@ def hard_delete(self):
+                    key = self._download_file_from_lcoal_storage(_url)
+            except FileNotFoundError:
+                pass
+        if not key:
+            for _url in urls:
+                try:
+                    if _url.startswith("http"):
+                        key = self._download_file_from_url(_url)
+                except ResourceNotFoundError:
+                    pass
+        return key
+
+    def _add_file(self, url, original_url=None, filename=None, **kwargs):
+        """Downloads file from url and saves it with `filename` as a proper name.
+        If filename is not provided ti will be resolved in the following way:
+            - Use filename if it's provided
+            - Check if key looks as proper filename, if yes then use it.
+            - if original_url is provided and last part of it (after last `/`)
+              looks like proper filename then use it
+            - as last resort use last part of `url` (after last `/`)
+
+        Args:
+            url (string): Url to a file, or to local api
+            filename (string): Proper name of the file.
+            original_url (string): URL from which this document was downloaded
+        Keyword Args:
+            description (string): works for documents and figures
+            fulltext (bool): works for documents only
+            hidden (bool): works for documents only
+            material (string): works for documents and figures
+            caption (string): works for facets only
+            key (string): Can contain name of the file (compatibility with inspire-next)
+
+        Returns (dict): Metadata for file
+
+        """
+
+        def is_filename(test_str):
+            """Checks if provided string looks like proper filename
+
+            Args:
+                test_str: String for testing
+
+            Returns: True if filename looks valid, False otherwise.
+
+            """
+            match = re.match(
+                "^([a-zA-Z0-9_]+)\.(?!\.)([a-zA-Z0-9]{1,5})(?<!\.)$", test_str
+            )
+            if match and match.start() == 0:
+                return True
+            return False
+
+        key = self._find_and_add_file(url, original_url)
+        if not key:
+            raise FileNotFoundError(f"File `{url}|{original_url}` not found")
+
+        if not filename:
+            _original_key = kwargs.get("key")
+            if key and is_filename(key):
+                filename = key
+            if not filename and original_url:
+                filename = original_url.split("/")[-1]
+                if not is_filename(filename):
+                    filename = None
+            if not filename:
+                filename = url.split("/")[-1]
+
+        metadata = kwargs
+        metadata["key"] = key
+
+        if not original_url:
+            metadata["original_url"] = url
+        else:
+            metadata["original_url"] = original_url
+
+        metadata["filename"] = filename
```

Comment:
filename can be ``''`` here no? https://github.com/inspirehep/inspirehep/pull/5/files#diff-b05b21cfe41c0c87d42e71a8841ba419R386

Reply:
Filename shouldn't be empty. In schemas I allowed it to be empty for back compatibility with inspire-next, but for inspirehep it shouldn't happen so in worst case scenario (filename not provided, not in key and not in url) filename will be same as key.